### PR TITLE
redux-loop compatibility

### PR DESCRIPTION
**Pull request by rikukissa**
[redux-loop](https://github.com/redux-loop/redux-loop) is quite popular library being used in conjunction with Redux. It seems to be working nicely with redux-subspace, except for the fact that the side effects returned by the reducers do not get namespaced like they should.

Basically this library would be a wrapper around the default `namespaced` higher-order reducer that takes in the "loop" returned by the reducer and modifies the command so that all the actions included will be namespaced.

This PR is still WIP and is lacking documentation

Comments are welcome :)

https://github.com/ioof-holdings/redux-subspace/issues/38

### CODE REVIEW

**Comment by jpeyper** [comment]
Thanks for your PR.

We will take a look at this (and redux-loop) over the next few days :)

One question, you mentioned compatiblility with redux-loop v2. Do you feel this is important? Do you think it will be difficult to have compatibility with both within the same package?

My personal thoughts is to support the latest version only.

**Comment by rikukissa** [comment]
I would go with supporting only the latest version as well, but unfortunately I need the v2 support myself 😔 The API changes they've made between v2 and v3 have been quite drastic and since I'm currently working on a fairly large codebase, upgrading isn't currently an option.  

Support for the previous version is easy to implement and I'd argue it would be a good thing for this library, especially from the easy adoption point of view. I'm guessing there's a lot of projects still stuck with v2.

**Comment by rikukissa** [comment]
I have to take my word back about it being easy to support both versions at the same time. Turned out to be much more difficult than I thought. Lets only support the latest version like you suggested. I'll do fork for v2.0 at some point.

**Comment by mpeyper** [comment]
Can you branch the logic by the available redux-loop functions?

Something like

```
import * as ReduxLoop from 'redux-loop'

if (ReduxLoop.getCmd && ReduxLoop.getEffect) {
	// version 3
	// handle either
} else if (ReduxLoop.getEffect) {
	// version 2
	// only handle effects
} else {
	// version 4 (or whenever effects get removed)
	// only handle commands
}
```

As for publishing a forked version, you could try something like [what redux-loop suggests](https://github.com/redux-loop/redux-loop#dont-see-a-feature-you-want).

**Comment by rikukissa** [comment]
Tried doing that, and it would actually work. The problem that I had was related to testing whether our logic is correct for handling the effects created by redux-loop v2. In our unit tests I would have to manually create the v2 style effects, since redux-loop's v3 API doesn't provide a way for generating them. They do have a module called [effects.js](https://github.com/redux-loop/redux-loop/blob/master/modules/effects.js) for backwards compatibility reasons, but it's basically a wrapper around the new command system. 

Now the problem with creating them manually is that our [isLoop call](https://github.com/ioof-holdings/redux-subspace/pull/44/files#diff-6d98d47b7281e2611bec439e9fff2673R49) starts returning undefined because the [logic behind it](https://github.com/redux-loop/redux-loop/blob/master/modules/cmd.js#L19-L21) relies on a symbol that isn't exported from redux-loop at all.

So the options I see here are
  - Implement the v2 support without unit tests
  - Publish v2 version as a separate package. Here we could install v2 as a devdep and use it in tests

Thanks for the link 👍  That would work indeed. Didn't even think of that to be honest.

**Comment by rikukissa** [comment]
Hey I think I figured out another solution! Maybe we could first create v3 style commands and modify them to look like v2 effects while still keeping the symbol key in the effect object. I'll try that and report back!

**Comment by rikukissa** [comment]
Hey, it worked! The only downsides of doing this is that we now get some warnings to our test output: 
image

and that I needed to create [this factory function](https://github.com/ioof-holdings/redux-subspace/pull/44/files#diff-6d98d47b7281e2611bec439e9fff2673R78) so I'm able to mimic the v2 environment in our tests.

Tried it in 2 of my projects running different version of redux-loop and it seems to be working fine 🙂

**Comment by mpeyper** [comment]
Great!  We can look at suppressing console output for those test cases, but I don't think it's necessary for this PR atm.

File: packages/redux-subspace-loop/package.json [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/ioof-holdings/redux-subspace.git"
+  },
+  "dependencies": {
+    "redux-subspace": "^2.0.3-alpha"
+  },
+  "peerDependencies": {
+    "redux-loop": ">=2.0.0 <3.0.0"
```
Comment (mpeyper) [inline packages/redux-subspace-loop/package.json]: should this be `^2.0.0 || ^3.0.0`?

File: packages/redux-subspace-loop/test/integration-spec.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { createStore } from 'redux'
+import { Cmd, loop, install, combineReducers } from 'redux-loop'
+import { namespaced } from '../src'
+import { subspace } from 'redux-subspace'
+
+describe('integration tests', () => {
```
Comment (mpeyper) [inline packages/redux-subspace-loop/test/integration-spec.js]: I would like to see `should work with namespaced single subspace` and `should work with namespaced nested subspaces` integration tests as well.

File: packages/redux-subspace-loop/package.json (hunk 1 above)
Comment (rikukissa) [reply packages/redux-subspace-loop/package.json]: Good point! Fixed.

File: packages/redux-subspace-loop/src/index.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        if (effect.type === 'BATCH') {
+            return {
+                ...effect,
+                effects: effect.effects.map(namespaceEffect)
+            }
+        }
+
+        return effect
```
Comment (mpeyper) [inline packages/redux-subspace-loop/src/index.js]: code coverage shows this line is uncovered (not super important, just FYI).

File: packages/redux-subspace-loop/docs/Usage.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# Usage
+
+Using Redux Subspace with [`redux-loop`](https://github.com/redux-loop/redux-loop) is pretty straightforward and should require no additional enhancer or middleware configuration. The only difference is that you have to import the `namespaced` function from `redux-subspace-loop` instead using the default one redux-subspace package. This extends the functionality provided by redux-subspace to also namespace the commands returned by your reducer.
```
Comment (mpeyper) [inline packages/redux-subspace-loop/docs/Usage.md]: > The only difference is that you have to import the `namespaced` function from `redux-subspace-loop` instead **_of_** using the default one **_from_** the redux-subspace package

Just a suggestion, feel free to reword your own way.

File: packages/redux-subspace-loop/docs/api/namespaced.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# `namespaced(namespace)`
+
+A higher-order reducer that filters out actions that do not match the provided namespace and namespace the commands returned by your reducer.
```
Comment (mpeyper) [inline packages/redux-subspace-loop/docs/api/namespaced.md]: > A higher-order reducer that filters out actions that do not match the provided namespace and **_namespaces_** the commands returned by your reducer.

Again, just a suggestion

**Comment by mpeyper** [comment]
I've looked through the code and the approach, naming and result is consistent with redux-subspace and the other compatibility packages.  Good job!

There should probably be a page added in the main docs similar to [the middleware page](https://github.com/ioof-holdings/redux-subspace/blob/master/docs/advanced/middleware/README.md) about compatible enhancers.  I'm not asking you to do this (unless you want to), but it should be done shortly after this is merged.

**Comment by mpeyper** [comment]
@rikukissa Let me know if you want to make the suggested changes yourself before I merge this, or if you want me to merge it and someone (probably me) can make them later.

Either way, I'm happy for this to be merged.

File: packages/redux-subspace-loop/test/integration-spec.js (hunk 2 above)
Comment (rikukissa) [reply packages/redux-subspace-loop/test/integration-spec.js]: Will add them today :)

**Comment by rikukissa** [comment]
Thanks for the review! I'll make some changes tonight so we can get this merged 👍

**Comment by rikukissa** [comment]
Thanks again for the review @mpeyper!  Addressed all the things you mentioned and from my perspective, this PR should now be good to go. Might make sense to take a quick look at the commits I pushed tho 🙂

**Comment by rikukissa** [comment]
Suppressed the redux-loop effect nags :)

image

**Comment by mpeyper** [comment]
Looks good.

**Pull request merged**